### PR TITLE
Add shape inference support for a few training ops

### PR DESCRIPTION
**Pull request by alberto-magni**
**Description**: Describe your changes.

Add shape inference support for the following ops:
* SoftmaxCrossEntropy
* SoftmaxCrossEntropyLossGrad
* SoftmaxCrossEntropyGrad
* LayerNormalizationGrad

**Motivation and Context**
- Why is this change required? What problem does it solve?

The goal is to be able to propagate shape information to all 
the nodes of the backward graph.

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by alberto-magni** [comment]
I have tried to add unit tests for this change by refactoring a ShapeInference test helper class.
Here is the change:
https://github.com/alberto-magni/onnxruntime/commit/af2006c02a07bff1e498f71ced63ab3dd5712eaa

The problem with adding tests is that the training ops in the domain kMSDomain are not registered automatically, leading to errors like:
`Graph resolve failed: This is an invalid model. Error in Node:node0 : No Op registered for LayerNormalizationGrad with domain_version of 12`
I went around this problem by moving the modified ops to the kOnnxDomain (just as an experiment).

Please let me know if you see a way to fix this issue for good. In that case I can update this PR to also include unit tests.
Thanks

**Comment by alberto-magni** [comment]
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=251138&view=logs&j=91d18614-091d-543c-f7ca-c2dad642e5d4&t=692f94a2-0d1d-567b-a903-2e87034da53f
The build fails because of a docker issue: 
```
Unable to find image '***/internal/azureml/onnxruntimecentoscpubuild:ch9m' locally
docker: Error response from daemon: manifest for ***/internal/azureml/onnxruntimecentoscpubuild:ch9m not found: manifest unknown: manifest tagged by "ch9m" is not found.
See 'docker run --help'.
```

**Review by SherlockNoMad — DISMISSED** [review]
:shipit:

**Comment by SherlockNoMad** [comment]
FYI, we are moving to use the onnx version of softmaxcrossentropy
https://github.com/onnx/onnx/blob/master/docs/Operators.md#softmaxcrossentropyloss

the internal version of the loss functions will be deprecated.

**Comment by alberto-magni** [comment]
Thanks.
Good to know about the upcoming deprecation. I was confused by the duplication.
Can you please trigger the build again ? I don't have permissions to do so.

**Comment by thiagocrepaldi** [comment]
> Thanks.
> Good to know about the upcoming deprecation. I was confused by the duplication.
> Can you please trigger the build again ? I don't have permissions to do so.

CI triggered!

**Comment by alberto-magni** [comment]
Hi all,
I see a few CI test failures:
https://github.com/microsoft/onnxruntime/pull/5660/checks?check_run_id=1359822004
* ```
Unable to find image '***/internal/azureml/onnxruntimecentoscpubuild:ch9m' locally
docker: Error response from daemon: manifest for ***/internal/azureml/onnxruntimecentoscpubuild:ch9m not found: manifest unknown: manifest tagged by "ch9m" is not found.
See 'docker run --help'.
```
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=251138&view=logs&j=c0b799d6-18b6-56f3-32a9-d967e9a2810f&t=76b5f847-1321-53a1-b83a-0b79284f9300
* ReduceSum_MidTensor: if a flaky test
* Run/507 ... and similar fail with this error:
`[ONNXRuntimeError] : 1 : FAIL : Non-zero status code returned while running Conv node. Name:'convolution' Status Message: CUDNN error executing cudnnConvolutionForward(CudnnHandle(), &alpha, s_.x_tensor, x_data, s_.filter_desc, w_data, s_.conv_desc, s_.algo, workspace.get(), s_.workspace_bytes, &beta, s_.y_tensor, y_data)`
I cannot repro these tests locally on my machine because the repository does not include the onnx files needed by these tests: 
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/test/providers/cpu/model_tests.cc
I can investigate the issue if you point me to where to get the test data.

**Review by tlh20 — APPROVED** [review]
